modules/streams/src/main/scala/dev/profunktor/redis4cats/pubsub/internals/PubSubInternals.scala

```diff
@@ -46,14 +42,14 @@ object PubSubInternals {
   private[redis4cats] def apply[F[_]: ConcurrentEffect: Log, K, V](
       state: Ref[F, PubSubState[F, K, V]],
       subConnection: StatefulRedisPubSubConnection[K, V]
-  ): GetOrCreateTopicListener[F, K, V] = { channel => st =>
-    st.get(channel.underlying)
+  ): GetOrCreateTopicListener[F, K, V] = { channels => st =>
+    st.get(channels.head.underlying)
```

Comment:
This will create only a topic and listener for the first channel in the sequence, no? Generally, using `head` is discouraged.

Reply:
I agree .head is wrong, it was done just to satisfy the compiler, this PR was more of a conversation starter.
My question is, why do we need to keep track of the subscribed channels? 
I'm sure redis doesn't care if we submit a double unsubscribe.
I the state tracking is removed, this line will go way.
I'll have a go later, and I'll add some tests.

Reply:
> why do we need to keep track of the subscribed channels?

I believe tests will answer your question but I guess it's because we need to handle internal `Topics` that are mapped to the underlying Java listener. 

> I'm sure redis doesn't care if we submit a double unsubscribe.

Redis won't probably care but at call site we need to make sure `unsubscribe` actually works (e.g. we no longer receive messages, easy to verify with a test).

Reply:
Had some time today to look at this. I've added some tests for pub/sub as there were none. 
I've changed the way we subscribe to Lettuce, it now adds one listener per connection, and the listener will route messages 
to the appropriate Topic based on the Map in state. Previously it was adding a new Listener per Topic, but Lettuce sends all messages to all Listeners (it just iterates through a Listener List), so multiple Listeners are not necessary.
I've also made `Subscriber` a `Resource` and it sends UNSUBSCRIBE to Redis when the connection is closed.